rest/route/aws.go

```diff
@@ -98,3 +108,98 @@ func (aws *awsSns) Run(ctx context.Context) gimlet.Responder {
 }
+
+type eventBridgeNotification struct {
+	DetailType string      `json:"detail-type"`
+	Detail     eventDetail `json:"detail"`
+}
+type eventDetail struct {
+	InstanceID string `json:"instance-id"`
+	State      string `json:"state"`
+}
+
+func (aws *awsSns) handleNotification(ctx context.Context) gimlet.Responder {
+	notification := &eventBridgeNotification{}
+	if err := json.Unmarshal([]byte(aws.payload.Message), notification); err != nil {
+		return gimlet.NewJSONErrorResponse(errors.Wrap(err, "problem unmarshalling notification"))
+	}
+
+	switch notification.DetailType {
+	case interruptionWarningType:
+		if err := aws.handleInstanceInterruptionWarning(ctx, notification.Detail.InstanceID); err != nil {
+			return gimlet.NewJSONInternalErrorResponse(errors.Wrap(err, "problem processing interruption warning"))
+		}
+
+	case instanceStateChangeType:
+		if notification.Detail.State == ec2.InstanceStateNameTerminated {
+			if err := aws.handleInstanceTerminated(ctx, notification.Detail.InstanceID); err != nil {
+				return gimlet.NewJSONInternalErrorResponse(errors.Wrap(err, "problem processing instance termination"))
+			}
+		}
+
+	default:
+		return gimlet.NewJSONErrorResponse(errors.Errorf("unknown detail type '%s'", notification.DetailType))
+	}
+
+	return gimlet.NewJSONResponse(struct{}{})
+}
+
+func (aws *awsSns) handleInstanceInterruptionWarning(ctx context.Context, instanceID string) error {
+	h, err := aws.sc.FindHostById(instanceID)
+	if err != nil {
+		return err
+	}
+	// don't make AWS keep retrying if we haven't heard of the host
+	if h == nil {
+		return nil
+	}
+
+	if h.Status == evergreen.HostTerminated {
+		return nil
+	}
+
+	if skipEarlyTermination(h) {
+		return nil
+	}
+
+	// return success on duplicate job errors so AWS won't keep retrying
+	_ = aws.queue.Put(ctx, units.NewHostTerminationJob(aws.env, h, true, "got interruption warning"))
+
+	return nil
+}
+
+// skipEarlyTermination decides if we should terminate the instance now or wait for AWS to do it.
+// See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-interruptions.html#billing-for-interrupted-spot-instances
+func skipEarlyTermination(h *host.Host) bool {
+	// Let AWS terminate if we're within the first hour
+	if utility.IsZeroTime(h.StartTime) || time.Now().Sub(h.StartTime) < time.Hour {
+		return true
+	}
+
+	// Let AWS terminate Windows hosts themselves.
+	if h.Distro.IsWindows() {
+		return true
+	}
+
+	return false
+}
+
+func (aws *awsSns) handleInstanceTerminated(ctx context.Context, instanceID string) error {
+	h, err := aws.sc.FindHostById(instanceID)
+	if err != nil {
+		return err
+	}
+	// don't make AWS keep retrying if we haven't heard of the host
+	if h == nil {
+		return nil
+	}
+
+	if h.Status == evergreen.HostTerminated {
+		return nil
+	}
+
+	// return success on duplicate job errors so AWS won't keep retrying
+	_ = aws.queue.Put(ctx, units.NewHostMonitorExternalStateJob(aws.env, h, aws.payload.MessageId))
```

Comment:
Curiosity question: is this queue used for specific things? I know a lot of routes use env.RemoteQueue()

Reply:
the remote queue is plumbed through:
[here](https://github.com/evergreen-ci/evergreen/blob/89399f0dff06928962b877a93c2e181bd8ffd4fc/operations/service_web.go#L52) -> [here](https://github.com/evergreen-ci/evergreen/blob/89399f0dff06928962b877a93c2e181bd8ffd4fc/operations/service_web.go#L80) -> [here](https://github.com/evergreen-ci/evergreen/blob/89399f0dff06928962b877a93c2e181bd8ffd4fc/operations/service_web.go#L190) stores the queue in the API server.
The queue is [passed](https://github.com/evergreen-ci/evergreen/blob/89399f0dff06928962b877a93c2e181bd8ffd4fc/service/service.go#L58) to the routes.